<a href="https://colab.research.google.com/github/Vira2021/Orbit_Future_Academi/blob/main/Copy_of_simple_data_pre_processing_steps_(1).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Data Pre-processing Stage

 
  Pada notebook ini berisi tentang basic dalam data pre-processing


#### Why preprocessing ?
1. Data yang kita punya dalam dunia nyata biasanya:
    * Incomplete: lacking attribute values, lacking certain attributes of interest, or containing only aggregate data.
    * Noisy: mengandung error seperti missing values dan outliers.
    * Inconsistent: mengandung perbedaan dalam penulisan kolom.

Mari kita pakai sebuah dataset sederhana untuk latihan.
Dataset "data.csv" mengandung informasi dimana apakah user akan membeli sebuah product atau tidak.
User mempunya fitur age,salary, dan negara tempat mereka tinggal.

## 1. Import library

Pada percobaan kali ini kita hanya membutuhkan 2 library yaitu pandas dan numpy. Pandas berguna untuk mengimport datasetnya sedangkan numpy akan kita gunakan dalam mempreproses datanya.

In [ ]:
import numpy as np
import pandas as pd

## 2 Import dataset

In [ ]:
dataset = pd.read_csv("Data.csv")


# Menampilkan shape dari data yg kita punya
print('dataset:', (str(dataset.shape)))


dataset: (75, 4)


Dataset kita mengandung 15 rows dan 4 columns.
Itu artinya kita mempunyai 15 data dan 4 Kolom

In [ ]:
#print the dataset
dataset


Country  Age  Salary Purchased
0          India   34   92000       Yes
1      Sri lanka   22   89700       Yes
2          China   31   74000       Yes
3      Sri lanka   29   78900       Yes
4          China   55   98000       Yes
..           ...  ...     ...       ...
70        Komoro   56   89065       Yes
71         Kosta   46   54329        No
72  Kroasia Kuba   65   54321       Yes
73       Lebanon   45   45670        No
74        Lesoto   49   76549        No

[75 rows x 4 columns]

In [ ]:
# Memisahkan antara fitur dan label

# Fitur
# iloc[rows,columns] (Slicing pada dataframe)
# Mengambil semua data dalam barisan.
# Mengambil kolom index ke-0 sampai ke-2
X = dataset.iloc[:,:3].values

# Label
# iloc[rows,columns] (Slicing pada dataframe)
# Mengambil semua data dalam barisan.
# Hanya Mengambil kolom index ke-3
Y = dataset.iloc[:,3].values

In [ ]:
# Print the X and Y
print ('Fitur (X):', X)
print ('\n-----------------------------------')
print ('\nLabel (Y): ',Y)

Fitur (X): [['India' 34 92000]
 ['Sri lanka' 22 89700]
 ['China' 31 74000]
 ['Sri lanka' 29 78900]
 ['China' 55 98000]
 ['India' 24 56790]
 ['Sri lanka' 28 90790]
 ['India' 50 60000]
 ['China' 51 89000]
 ['India' 44 78000]
 ['Sri lanka' 21 98000]
 ['China' 25 98087]
 ['India' 33 45000]
 ['India' 42 65000]
 ['Sri lanka' 33 22000]
 ['India' 34 92000]
 ['Sri lanka' 22 25000]
 ['China' 31 74000]
 ['Sri lanka' 29 78900]
 ['India' 50 60000]
 ['China' 51 89000]
 ['India' 44 78000]
 ['Sri lanka' 21 87099]
 ['China' 25 30000]
 ['India' 33 45000]
 ['India' 42 65000]
 ['Sri lanka' 33 22000]
 ['Sri lanka' 29 78900]
 ['Jawa barat' 34 92000]
 ['Jawa Tengah' 22 90700]
 ['Jawa Timur' 31 74000]
 ['Bekasi' 29 78900]
 ['Australia' 55 98000]
 ['Singapura' 24 54000]
 ['Medan' 28 98000]
 ['Makasar' 50 60000]
 ['Arab saudi' 51 89000]
 ['Bahama' 44 78000]
 ['Austria' 21 56900]
 ['Barbandos' 25 89070]
 ['Belanda' 33 97800]
 ['Brasil' 42 65000]
 ['Korea' 33 98609]
 ['Belgia ' 34 92000]
 ['Belise' 22 96900]
 ['B

## 3. Handle Missing Data


In [ ]:
# isnull : berguna untuk memeriksa suatu kolom ada datanya atau tidak.
dataset.isnull().sum()

Country      0
Age          0
Salary       0
Purchased    0
dtype: int64

Dengan menggunakan fungsi isnull().sum() kita dapat melihat jumlah data yg memiliki missing value dari tiap kolom yang ada.
Disini kita dapat melihat kalau terdapat 2 kolom yang memiliki masalah missing value, yaitu pada kolom 'Age' dan 'Salary'.

Untuk menangangi masalah missing value, kita bisa melakukan 2 hal. Pertama kita bisa menghapus datanya atau kita bisa mengisi nilai yang kosong itu menggunakan nilai mean/modus/median.

Pada percobaan kali ini kita akan menggunakan nilai Mean untuk mengisi missing value tersebut menggunakan bantuan Simple Imputer dari module sklearn

In [ ]:
from sklearn.impute import SimpleImputer

# SimpleImputer class mempunyai 2 parameter, yaitu :
#     missing_values : Missing value pada data kita, biasanya diisi dengan np.nan
#     strategy       : Strategi yang akan kita gunakan dalam menangani missing value (Mean/Modus/Median).
#                      Defaultnya parameter strategy 'mean'

imputer = SimpleImputer(missing_values = np.nan ,strategy = 'mean')


# Fit imputer kedalam data X untuk kolom Age (index ke-1) dan Salary (index ke-2).
imputer = imputer.fit(X[:,1:3])

#Replace missing value dengan mean pada kolom tersebut
X[:,1:3] = imputer.transform(X[:,1:3])


Dibawah ini kita bisa melihat bahwa tidak ada nilai 'Nan' pada data X

In [ ]:
print ('X: ', X)

X:  [['India' 34.0 92000.0]
 ['Sri lanka' 22.0 89700.0]
 ['China' 31.0 74000.0]
 ['Sri lanka' 29.0 78900.0]
 ['China' 55.0 98000.0]
 ['India' 24.0 56790.0]
 ['Sri lanka' 28.0 90790.0]
 ['India' 50.0 60000.0]
 ['China' 51.0 89000.0]
 ['India' 44.0 78000.0]
 ['Sri lanka' 21.0 98000.0]
 ['China' 25.0 98087.0]
 ['India' 33.0 45000.0]
 ['India' 42.0 65000.0]
 ['Sri lanka' 33.0 22000.0]
 ['India' 34.0 92000.0]
 ['Sri lanka' 22.0 25000.0]
 ['China' 31.0 74000.0]
 ['Sri lanka' 29.0 78900.0]
 ['India' 50.0 60000.0]
 ['China' 51.0 89000.0]
 ['India' 44.0 78000.0]
 ['Sri lanka' 21.0 87099.0]
 ['China' 25.0 30000.0]
 ['India' 33.0 45000.0]
 ['India' 42.0 65000.0]
 ['Sri lanka' 33.0 22000.0]
 ['Sri lanka' 29.0 78900.0]
 ['Jawa barat' 34.0 92000.0]
 ['Jawa Tengah' 22.0 90700.0]
 ['Jawa Timur' 31.0 74000.0]
 ['Bekasi' 29.0 78900.0]
 ['Australia' 55.0 98000.0]
 ['Singapura' 24.0 54000.0]
 ['Medan' 28.0 98000.0]
 ['Makasar' 50.0 60000.0]
 ['Arab saudi' 51.0 89000.0]
 ['Bahama' 44.0 78000.0]
 ['Austria'

* Mean Value of Age    = Sum of all age values /14   = 33.714285714285715
* Mean Value of Salary = Sum of all Salary value /14 = 54857.142857142855

## 3. Encode the Categorical data

Categorical data biasanya biasanya memiliki tipe data string.
Beberapa algoritma dapat memproses categorical data yg berbentuka string secara langsung seperti decision tree.

Akan tetapi banyak machine learning algorithms tidak dapat memproses categorical data yg berbentuk string secara langsung. Mereka membutuhkan variabel yg berbentuk numeric.

Itu artinya categorical data harus dikonversi bentukanya kedalam numerical form.

Pada Dataset kita terdapat dua kolom categorical data

Yaitu kolom 'Country' dan 'Purchased'

#### Label Encoder: 

    * Metode ini bisa digunakan untuk mentranform non-numerical labels kedalam bentuk numerical labels (or nominal
      categorical variables).
    * Numerical labels akan selalu ada diantara 0 sampai n_classes-1. 
      Misalkan ada 3 kelas, berarti labelnya ada di 0, 1, dan 2.

In [ ]:
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
#Encode untuk data kategorikal pada kolom 'Country'
labelencoder_X = LabelEncoder()
X[:,0] = labelencoder_X.fit_transform(X[:,0])
X[:,0]

array([24, 49, 13, 49, 13, 24, 49, 24, 13, 24, 49, 13, 24, 24, 49, 24, 49,
       13, 49, 24, 13, 24, 49, 13, 24, 24, 49, 49, 33, 31, 32, 5, 1, 48,
       47, 46, 0, 3, 2, 4, 6, 9, 41, 7, 8, 10, 11, 12, 14, 16, 17, 15, 18,
       20, 19, 21, 22, 23, 25, 26, 27, 28, 29, 30, 34, 37, 36, 35, 38, 39,
       40, 42, 43, 44, 45], dtype=object)

Sekarang categorical value untuk kolom 'Country' dirbah ke dalam bentuk kategorikal

| Country | Value |
|:--------|:------|
| China   |   0   |  
| India   |   1   |   
| Srilanka|   2   |   


In [ ]:
X

array([[24, 34.0, 92000.0],
       [49, 22.0, 89700.0],
       [13, 31.0, 74000.0],
       [49, 29.0, 78900.0],
       [13, 55.0, 98000.0],
       [24, 24.0, 56790.0],
       [49, 28.0, 90790.0],
       [24, 50.0, 60000.0],
       [13, 51.0, 89000.0],
       [24, 44.0, 78000.0],
       [49, 21.0, 98000.0],
       [13, 25.0, 98087.0],
       [24, 33.0, 45000.0],
       [24, 42.0, 65000.0],
       [49, 33.0, 22000.0],
       [24, 34.0, 92000.0],
       [49, 22.0, 25000.0],
       [13, 31.0, 74000.0],
       [49, 29.0, 78900.0],
       [24, 50.0, 60000.0],
       [13, 51.0, 89000.0],
       [24, 44.0, 78000.0],
       [49, 21.0, 87099.0],
       [13, 25.0, 30000.0],
       [24, 33.0, 45000.0],
       [24, 42.0, 65000.0],
       [49, 33.0, 22000.0],
       [49, 29.0, 78900.0],
       [33, 34.0, 92000.0],
       [31, 22.0, 90700.0],
       [32, 31.0, 74000.0],
       [5, 29.0, 78900.0],
       [1, 55.0, 98000.0],
       [48, 24.0, 54000.0],
       [47, 28.0, 98000.0],
       [46, 50.0, 6000

In [ ]:
#Encode untuk data kategorikal pada kolom 'Purchased'
labelencoder_Y = LabelEncoder()
Y = labelencoder_X.fit_transform(Y)
Y

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 0, 1, 0, 1, 0, 0])

## 4. Splitting Dataset
train_test_split : salah satu metode yang dapat digunakan untuk mengevaluasi performa model machine learning. Metode evaluasi model ini membagi dataset menjadi dua bagian yakni bagian yang digunakan untuk training data dan untuk testing data dengan proporsi tertentu.

In [ ]:

from sklearn.model_selection import train_test_split

X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size = 0.2,random_state = 0)

In [ ]:
# Print the shape of the dataset
print ('X_train:', (str(X_train.shape)))
print ('----------------')
print ('X_test:', (str(X_test.shape)))
print ('----------------')
print ('Y_train:', (str(Y_train.shape)))
print ('----------------')
print ('Y_test:', (str(Y_test.shape)))
print ('----------------')

X_train: (60, 3)
----------------
X_test: (15, 3)
----------------
Y_train: (60,)
----------------
Y_test: (15,)
----------------


## 5. Scale your Features
 

Dalam sebuah dataset, biasanya nilai pada tiap fitur mempunyai range yg berbeda. Hal ini dapat menjadikan masalah untuk
beberapa model machine learning. Untuk itu kita harus melakukan scaling agar semua nilai pada fitur memiliki range yang sama. Pada kasus ini kita menggunakan StandardScaler
      
      

In [ ]:

from sklearn.preprocessing import StandardScaler

sc_X = StandardScaler()

# We need to fit and transform the training set
X_train = sc_X.fit_transform(X_train)

# We need to fit the test set
X_test = sc_X.transform(X_test)

In [ ]:
X_train

array([[ 0.98072968,  1.47199898,  0.72447956],
       [ 1.6001379 , -0.79261483,  0.25117857],
       [-0.67102557,  0.46550395,  0.20927292],
       [ 0.08602892,  0.54937854, -1.80731992],
       [-0.25808676,  0.96875146,  1.17347528],
       [ 1.6001379 , -0.87648942,  0.80479872],
       [-0.60220244, -1.46361152,  0.62832938],
       [ 0.84308341,  1.3042498 ,  1.10903371],
       [-0.87749498,  1.38812439,  1.14050951],
       [ 0.77426027, -0.87648942,  1.17338216],
       [-0.80867184,  1.05262605,  0.72145304],
       [-0.87749498, -0.62486566,  0.0230256 ],
       [-1.08396438, -0.62486566,  0.0230256 ],
       [ 0.22367519,  1.13650063,  0.65361246],
       [-0.87749498, -1.12811318,  1.14456039],
       [ 0.4301446 , -0.62486566,  0.0230256 ],
       [ 1.6001379 , -0.79261483,  0.25117857],
       [ 0.29249833,  1.13650063, -1.81304702],
       [ 1.6001379 , -1.46361152,  1.14050951],
       [-1.42808006, -0.79261483,  0.25117857],
       [-1.63454947, -1.46361152, -0.773

In [ ]:
X_test

array([[-0.5333793 , -0.45711649,  0.80526434],
       [-0.18926362,  1.47199898, -0.26146718],
       [ 1.6001379 , -1.46361152,  0.63293901],
       [ 1.32484536,  0.88487688,  0.1417117 ],
       [ 0.49896773, -0.37324191,  0.86113853],
       [ 1.6001379 , -0.45711649, -2.39818953],
       [ 0.70543714,  0.04613102,  0.14166514],
       [-0.46455616, -0.45711649,  1.14376884],
       [ 1.46249163, -0.87648942,  1.14050951],
       [-0.12044049,  0.96875146, -0.62884001],
       [-1.29043379, -0.37324191,  0.86113853],
       [ 0.01720578,  2.22687025, -1.29569853],
       [-1.15278752,  0.29775478, -0.39603086],
       [-1.35925692, -0.45711649,  1.13119714],
       [ 1.53131476, -1.21198776, -0.90821099]])

Nah Sekarang semua data dalam fitur sudah berada dalam range yg sama, data ini bisa lgsg kita masukan kedalam model machine learning.

Dalam percobaan kali ini kita akan mencoba memasukkan data kedalam model KNN.

KNN : sebuah metode klasifikasi terhadap sekumpulan data berdasarkan pembelajaran  data yang sudah terklasifikasikan sebelumya

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
k = 3
#Train Model and Predict  
knn = KNeighborsClassifier(n_neighbors = k).fit(X_train,Y_train)
knn

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=3, p=2,
                     weights='uniform')

In [ ]:
y_predict = knn.predict(X_test)

In [ ]:
from sklearn import metrics
print("Test set Accuracy: ", metrics.accuracy_score(Y_test, y_predict))

Test set Accuracy:  0.9333333333333333


Dalam hal ini kita hanya mendapat akurasi 93%, ini terhitung kecil. ini bisa disebabkan karena data yg kita punya hanyaa sedikit. 

**Thanks You**